### ISSUE TITLE
Separate cluster creation into two phase

### ISSUE THREAD

**Issue by saivirtue** [oharastream/ohara#1544]
We need to handle the cluster creation in a stage way :
1. store the request properties in configurator
2. create cluster

As #1942 discussed, the following issue should satisfy:
<details open><summary>requirement list</summary>

- create zookeeper data: POST with payload, response created object
- start zookeeper: PUT without payload, response httpcode=202
- stop zookeeper: PUT without payload, response httpcode=202
- delete zookeeper data: PUT without payload, response httpcode=204
- get zookeeper information: GET without payload, response up-to-date object (include status)

</details>

Since this issue change ALL cluster APIs, we separate to following tasks:
- [x] Separate cluster creation of zookeeper (#1869)
-- Zookeeper start should be idempotent (issue #1905)
- [x] Separate cluster creation of broker (#1971)
- [ ] Separate cluster creation of worker (#2144)
-- move auto-filled logic to creation of worker route (added from #2005)
- [x] ~Separate cluster creation of stream~ stream cluster is already two-phase creation
-- Integrate StreamAppDescription into StreamClusterInfo (issue #1025

**Comment by chia7712** [comment]
@saivirtue Auto-completing properties in START phase produces a embarrasses scenario that users can’t observer what is used to build their services. Hence, users will get a cluster they don’t like.

User is able to see the changes of properties if we auto-complete properties in CREATE phase, because the response is the properties we have updated.

Any feedbacks?

**Comment by saivirtue** [comment]
> Auto-completing properties in START phase produces a embarrasses scenario that users can’t observer what is used to build their services. Hence, users will get a cluster they don’t like.

We already add some description about auto-completed parameters in docs.
I think this is a trade-off situation: if you don't like so many parameters, you should care "the default" value.

**Comment by chia7712** [comment]
> if you don't like so many parameters, you should care "the default" value.

user CANNOT care the default value before starting if we auto-complete the properties in start phase And I don't think the docs is a "good" place to read the latest default value :(

**Comment by saivirtue** [comment]
> user CANNOT care the default value before starting if we auto-complete the properties in start phase 

Sure. User need to request another GET to get the value.
But as I worried about, if auto-completed was happened in creation phase:
step 1. absent parameter in creation -> response auto-completing
step 2. start this object, and then stop
step 3. update auto-completed parameter -> response should override the parameter?

If no, that is a weird response, so skip it.
if yes, we will override the data in configurator, and start action again will run in a different situation.

for example, 
```
> wk creation without broker
{
  "name": "wk01"
}

> start worker (it should try to find the default broker), and then stop.

> update broker name
{
  "broker": "some"
}

> start this worker again, the old data in default broker will be loss.
```

If I have miss something or misunderstanding, please tell me :)

**Comment by chia7712** [comment]
> start worker (it should try to find the default broker), and then stop.

the broker cluster should be auto-completed in Create phase. And the "auto-complete" should be for ignored fields only.

```
> wk creation without broker
{
  "name": "wk01"
}
```
the above response should be a auto-completed properties consisting of name, brokerClusterName and other configs.

> > start worker (it should try to find the default broker), and then stop.
> 
> > update broker name
> {
>   "broker": "some"
> }

It is ok to change the base cluster after stopping worker cluster. And the broker cluster is define so the "auto-complete" does not happen.

**Comment by saivirtue** [comment]
> It is ok to change the base cluster after stopping worker cluster. And the broker cluster is define so the "auto-complete" does not happen.

If so, I'm ok with moving the behavior happened in creation phase.
the rest concern for me is we need to do some "under works" besides stored data in configurator in POST...but it's ignorable, I guess.

**Comment by chia7712** [comment]
> "under works" 

Most of "default values" are constants so they are fine to us. By contrast, the default values which need to be produced by "under works" is tricky since they may produce exception. For that case,   we ought to just "ignore" the exception and then leave empty field. Finally, user observer the error when starting something with the uncompleted properties :)

**Comment by saivirtue** [comment]
Since I've announced the changes (auto complete in start) twice in daily meeting... 🤣 
We should open a new issue to apply this consensus and /cc to all developers.

**Comment by chia7712** [comment]
> We should open a new issue to apply this consensus and /cc to all developers.

Just file another issue to include the conclusion and Ping other guys to give feedbacks :)

**Comment by chia7712** [comment]
@saivirtue @jackyoh @wu87988622 could we remove the payload for all "action" APIs (start, stop, addNode, removeNode and so on) for 0.7?

**Comment by saivirtue** [comment]
> @saivirtue @jackyoh @wu87988622 could we remove the payload for all "action" APIs (start, stop, addNode, removeNode and so on) for 0.7?

You mean request or response?
for request, I think we almost there except worker.
for response, I realize that we have response the HttpCode.202/204 with some data already. So you want to remove the payload of response?

**Comment by chia7712** [comment]
> I realize that we have response the HttpCode.202/204 with some data already. So you want to remove the payload of response?

Yep. We should remove the payload from response for 202/204. I guess Ohara Manager still rely on the payload.

**Comment by saivirtue** [comment]
> Yep. We should remove the payload from response for 202/204. I guess Ohara Manager still rely on the payload.

Sure thing. I am OK with that.
If @oharastream/frontend were OK with that also, this will be done in the final PR.

**Comment by jackyoh** [comment]
+1

**Comment by konekoya** [comment]
> I guess Ohara Manager still rely on the payload.

Cannot be 100% sure about this, but it's okay to remove from 0.7 for me. It should be much work for this.

> If @oharastream/frontend were OK with that also, this will be done in the final PR.

+1 for this :)

**Comment by chia7712** [comment]
@konekoya @wu87988622 Do you ready to file a PR to avoid accessing the response of all action requests (start, stop, pause, resume and so on)? The PR can go first :)

**Comment by chia7712** [comment]
```scala
  // we need a "copy" of this class to change the state of cluster
  // TODO : after #1544, we should have a basic clone() function to accept (newNodeNames, state, error)
  def clone2(state: Option[String], error: Option[String]): ClusterInfo =
    throw new AssertionError("[TODO] You need to define this method by class")
```

@saivirtue what is purpose to above? I fail to find the usage of clone2 from our code base. Could I file a PR to remove it?

**Comment by saivirtue** [comment]
> @saivirtue what is purpose to above? I fail to find the usage of clone2 from our code base. Could I file a PR to remove it?

It should be removed in PR #2144

**Comment by chia7712** [comment]
> It should be removed in PR #2144

ok
